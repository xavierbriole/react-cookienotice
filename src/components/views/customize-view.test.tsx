import { act, render } from '@testing-library/react'
import { vi } from 'vitest'

import CustomizeView from './customize-view'

describe('CustomizeView', () => {
  beforeEach(() => {
    vi.resetAllMocks()
  })

  afterEach(() => {
    vi.useRealTimers()
  })

  describe('should render', () => {
    it('with services', () => {
      const { asFragment } = render(
        <CustomizeView
          customizeTitleLabel='customizeTitleLabel'
          services={[
            {
              name: 'service1Name',
              description: 'service1Description',
              code: 'service1Code',
              alwaysActive: true,
            },
            {
              name: 'service2Name',
              description: 'service2Description',
              code: 'service2Code',
              alwaysActive: false,
            },
          ]}
          onAcceptButtonClick={() => {}}
          acceptButtonLabel='acceptButtonLabel'
          onBackButtonClick={() => {}}
          backButtonLabel='backButtonLabel'
          alwaysActiveLabel='alwaysActiveLabel'
          customizeAcceptAllButtonLabel='customizeAcceptAllButtonLabel'
          onCustomizeAcceptAllButtonClick={() => {}}
          customizeAcceptAllTimeout={1000}
        />,
      )

      expect(asFragment()).toMatchSnapshot()
    })

    it('with services empty', () => {
      const { asFragment } = render(
        <CustomizeView
          customizeTitleLabel='customizeTitleLabel'
          services={[]}
          onAcceptButtonClick={() => {}}
          acceptButtonLabel='acceptButtonLabel'
          onBackButtonClick={() => {}}
          backButtonLabel='backButtonLabel'
          alwaysActiveLabel='alwaysActiveLabel'
          customizeAcceptAllButtonLabel='customizeAcceptAllButtonLabel'
          onCustomizeAcceptAllButtonClick={() => {}}
          customizeAcceptAllTimeout={1000}
        />,
      )

      expect(asFragment()).toMatchSnapshot()
    })
  })

  it('should handle checkbox change', () => {
    const { getByLabelText } = render(
      <CustomizeView
        customizeTitleLabel='customizeTitleLabel'
        services={[
          {
            name: 'service1Name',
            description: 'service1Description',
            code: 'service1Code',
            alwaysActive: true,
          },
          {
            name: 'service2Name',
            description: 'service2Description',
            code: 'service2Code',
            alwaysActive: false,
          },
          {
            name: 'service3Name',
            description: 'service3Description',
            code: 'service3Code',
            alwaysActive: false,
          },
        ]}
        onAcceptButtonClick={() => {}}
        acceptButtonLabel='acceptButtonLabel'
        onBackButtonClick={() => {}}
        backButtonLabel='backButtonLabel'
        alwaysActiveLabel='alwaysActiveLabel'
        customizeAcceptAllButtonLabel='customizeAcceptAllButtonLabel'
        onCustomizeAcceptAllButtonClick={() => {}}
        customizeAcceptAllTimeout={1000}
      />,
    )

    expect(getByLabelText('service1Code')).toBeChecked()
    expect(getByLabelText('service2Code')).not.toBeChecked()
    expect(getByLabelText('service3Code')).not.toBeChecked()

    act(() => {
      getByLabelText('service2Code').click()
    })

    expect(getByLabelText('service2Code')).toBeChecked()

    act(() => {
      getByLabelText('service3Code').click()
    })

    expect(getByLabelText('service3Code')).toBeChecked()

    act(() => {
      getByLabelText('service3Code').click()
    })

    expect(getByLabelText('service3Code')).not.toBeChecked()
  })

  describe('should handle accept button click', () => {
    it('when no service is selected', () => {
      const onAcceptButtonClick = vi.fn()

      const { getByText } = render(
        <CustomizeView
          customizeTitleLabel='customizeTitleLabel'
          services={[
            {
              name: 'service1Name',
              description: 'service1Description',
              code: 'service1Code',
              alwaysActive: false,
            },
            {
              name: 'service2Name',
              description: 'service2Description',
              code: 'service2Code',
              alwaysActive: false,
            },
          ]}
          onAcceptButtonClick={onAcceptButtonClick}
          acceptButtonLabel='acceptButtonLabel'
          onBackButtonClick={() => {}}
          backButtonLabel='backButtonLabel'
          alwaysActiveLabel='alwaysActiveLabel'
          customizeAcceptAllButtonLabel='customizeAcceptAllButtonLabel'
          onCustomizeAcceptAllButtonClick={() => {}}
          customizeAcceptAllTimeout={1000}
        />,
      )

      act(() => {
        getByText('acceptButtonLabel').click()
      })

      expect(onAcceptButtonClick).toHaveBeenCalledTimes(1)
      expect(onAcceptButtonClick).toHaveBeenCalledWith([])
    })

    it('when one always active service is selected', () => {
      const onAcceptButtonClick = vi.fn()

      const { getByText } = render(
        <CustomizeView
          customizeTitleLabel='customizeTitleLabel'
          services={[
            {
              name: 'service1Name',
              description: 'service1Description',
              code: 'service1Code',
              alwaysActive: true,
            },
            {
              name: 'service2Name',
              description: 'service2Description',
              code: 'service2Code',
              alwaysActive: false,
            },
          ]}
          onAcceptButtonClick={onAcceptButtonClick}
          acceptButtonLabel='acceptButtonLabel'
          onBackButtonClick={() => {}}
          backButtonLabel='backButtonLabel'
          alwaysActiveLabel='alwaysActiveLabel'
          customizeAcceptAllButtonLabel='customizeAcceptAllButtonLabel'
          onCustomizeAcceptAllButtonClick={() => {}}
          customizeAcceptAllTimeout={1000}
        />,
      )

      act(() => {
        getByText('acceptButtonLabel').click()
      })

      expect(onAcceptButtonClick).toHaveBeenCalledTimes(1)
      expect(onAcceptButtonClick).toHaveBeenCalledWith(['service1Code'])
    })

    it('when one service is selected', () => {
      const onAcceptButtonClick = vi.fn()

      const { getByText, getByLabelText } = render(
        <CustomizeView
          customizeTitleLabel='customizeTitleLabel'
          services={[
            {
              name: 'service1Name',
              description: 'service1Description',
              code: 'service1Code',
              alwaysActive: true,
            },
            {
              name: 'service2Name',
              description: 'service2Description',
              code: 'service2Code',
              alwaysActive: false,
            },
          ]}
          onAcceptButtonClick={onAcceptButtonClick}
          acceptButtonLabel='acceptButtonLabel'
          onBackButtonClick={() => {}}
          backButtonLabel='backButtonLabel'
          alwaysActiveLabel='alwaysActiveLabel'
          customizeAcceptAllButtonLabel='customizeAcceptAllButtonLabel'
          onCustomizeAcceptAllButtonClick={() => {}}
          customizeAcceptAllTimeout={1000}
        />,
      )

      act(() => {
        getByLabelText('service2Code').click()
      })

      act(() => {
        getByText('acceptButtonLabel').click()
      })

      expect(onAcceptButtonClick).toHaveBeenCalledTimes(1)
      expect(onAcceptButtonClick).toHaveBeenCalledWith([
        'service2Code',
        'service1Code',
      ])
    })

    it('when multiple services are selected', () => {
      const onAcceptButtonClick = vi.fn()

      const { getByText, getByLabelText } = render(
        <CustomizeView
          customizeTitleLabel='customizeTitleLabel'
          services={[
            {
              name: 'service1Name',
              description: 'service1Description',
              code: 'service1Code',
              alwaysActive: true,
            },
            {
              name: 'service2Name',
              description: 'service2Description',
              code: 'service2Code',
              alwaysActive: false,
            },
            {
              name: 'service3Name',
              description: 'service3Description',
              code: 'service3Code',
              alwaysActive: false,
            },
          ]}
          onAcceptButtonClick={onAcceptButtonClick}
          acceptButtonLabel='acceptButtonLabel'
          onBackButtonClick={() => {}}
          backButtonLabel='backButtonLabel'
          alwaysActiveLabel='alwaysActiveLabel'
          customizeAcceptAllButtonLabel='customizeAcceptAllButtonLabel'
          onCustomizeAcceptAllButtonClick={() => {}}
          customizeAcceptAllTimeout={1000}
        />,
      )

      act(() => {
        getByLabelText('service2Code').click()
      })

      act(() => {
        getByLabelText('service3Code').click()
      })

      act(() => {
        getByText('acceptButtonLabel').click()
      })

      expect(onAcceptButtonClick).toHaveBeenCalledTimes(1)
      expect(onAcceptButtonClick).toHaveBeenCalledWith([
        'service2Code',
        'service3Code',
        'service1Code',
      ])
    })
  })

  it('should handle accept all button click', () => {
    vi.useFakeTimers()

    const onCustomizeAcceptAllButtonClick = vi.fn()

    const { getByText, getByLabelText } = render(
      <CustomizeView
        customizeTitleLabel='customizeTitleLabel'
        services={[
          {
            name: 'service1Name',
            description: 'service1Description',
            code: 'service1Code',
            alwaysActive: true,
          },
          {
            name: 'service2Name',
            description: 'service2Description',
            code: 'service2Code',
            alwaysActive: false,
          },
          {
            name: 'service3Name',
            description: 'service3Description',
            code: 'service3Code',
            alwaysActive: false,
          },
        ]}
        onAcceptButtonClick={() => {}}
        acceptButtonLabel='acceptButtonLabel'
        onBackButtonClick={() => {}}
        backButtonLabel='backButtonLabel'
        alwaysActiveLabel='alwaysActiveLabel'
        customizeAcceptAllButtonLabel='customizeAcceptAllButtonLabel'
        onCustomizeAcceptAllButtonClick={onCustomizeAcceptAllButtonClick}
        customizeAcceptAllTimeout={1000}
      />,
    )

    expect(getByLabelText('service1Code')).toBeChecked()
    expect(getByLabelText('service2Code')).not.toBeChecked()
    expect(getByLabelText('service3Code')).not.toBeChecked()

    act(() => {
      getByText('customizeAcceptAllButtonLabel').click()
      vi.runAllTimers()
    })

    expect(getByLabelText('service1Code')).toBeChecked()
    expect(getByLabelText('service2Code')).toBeChecked()
    expect(getByLabelText('service3Code')).toBeChecked()

    expect(onCustomizeAcceptAllButtonClick).toHaveBeenCalledTimes(1)
    expect(onCustomizeAcceptAllButtonClick).toHaveBeenCalledWith([
      'service1Code',
      'service2Code',
      'service3Code',
    ])
  })

  it('should handle back button click', () => {
    const onBackButtonClick = vi.fn()

    const { getByText } = render(
      <CustomizeView
        customizeTitleLabel='customizeTitleLabel'
        services={[
          {
            name: 'service1Name',
            description: 'service1Description',
            code: 'service1Code',
            alwaysActive: true,
          },
          {
            name: 'service2Name',
            description: 'service2Description',
            code: 'service2Code',
            alwaysActive: false,
          },
        ]}
        onAcceptButtonClick={() => {}}
        acceptButtonLabel='acceptButtonLabel'
        onBackButtonClick={onBackButtonClick}
        backButtonLabel='backButtonLabel'
        alwaysActiveLabel='alwaysActiveLabel'
        customizeAcceptAllButtonLabel='customizeAcceptAllButtonLabel'
        onCustomizeAcceptAllButtonClick={() => {}}
        customizeAcceptAllTimeout={1000}
      />,
    )

    act(() => {
      getByText('backButtonLabel').click()
    })

    expect(onBackButtonClick).toHaveBeenCalledTimes(1)
  })
})
