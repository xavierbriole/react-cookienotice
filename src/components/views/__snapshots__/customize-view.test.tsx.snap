// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CustomizeView > should render > with services 1`] = `
<DocumentFragment>
  <div
    class="react-cookienotice-body"
  >
    <div
      class="react-cookienotice-title"
    >
      <span
        class="react-cookienotice-label"
      >
        customizeTitleLabel
      </span>
    </div>
    <div
      class="react-cookienotice-services"
    >
      <div
        class="react-cookienotice-service"
      >
        <div
          class="react-cookienotice-service-checkbox"
        >
          <input
            checked=""
            disabled=""
            id="service1Code"
            name="service1Code"
            type="checkbox"
          />
          <label
            for="service1Code"
          >
            service1Code
          </label>
        </div>
        <div
          class="react-cookienotice-service-info"
        >
          <button
            class="react-cookienotice-service-info-title"
          >
            service1Name
            <div
              class="react-cookienotice-service-info-title-always-active"
            >
              <span
                class="react-cookienotice-label"
              >
                alwaysActiveLabel
              </span>
            </div>
          </button>
          <div
            class="react-cookienotice-service-info-content"
            style="max-height: 0;"
          >
            <span
              class="react-cookienotice-label"
            >
              service1Description
            </span>
          </div>
        </div>
      </div>
      <div
        class="react-cookienotice-service"
      >
        <div
          class="react-cookienotice-service-checkbox"
        >
          <input
            id="service2Code"
            name="service2Code"
            type="checkbox"
          />
          <label
            for="service2Code"
          >
            service2Code
          </label>
        </div>
        <div
          class="react-cookienotice-service-info"
        >
          <button
            class="react-cookienotice-service-info-title"
          >
            service2Name
          </button>
          <div
            class="react-cookienotice-service-info-content"
            style="max-height: 0;"
          >
            <span
              class="react-cookienotice-label"
            >
              service2Description
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="react-cookienotice-buttons"
  >
    <button
      class="react-cookienotice-button"
    >
      backButtonLabel
    </button>
    <button
      class="react-cookienotice-button"
    >
      customizeAcceptAllButtonLabel
    </button>
    <button
      class="react-cookienotice-button"
    >
      acceptButtonLabel
    </button>
  </div>
</DocumentFragment>
`;

exports[`CustomizeView > should render > with services empty 1`] = `
<DocumentFragment>
  <div
    class="react-cookienotice-body"
  >
    <div
      class="react-cookienotice-title"
    >
      <span
        class="react-cookienotice-label"
      >
        customizeTitleLabel
      </span>
    </div>
    <div
      class="react-cookienotice-services"
    />
  </div>
  <div
    class="react-cookienotice-buttons"
  >
    <button
      class="react-cookienotice-button"
    >
      backButtonLabel
    </button>
    <button
      class="react-cookienotice-button"
    >
      customizeAcceptAllButtonLabel
    </button>
    <button
      class="react-cookienotice-button"
    >
      acceptButtonLabel
    </button>
  </div>
</DocumentFragment>
`;
